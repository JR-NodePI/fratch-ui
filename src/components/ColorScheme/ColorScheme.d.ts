import { COLOR_SCHEMES } from './colorSchemeConstants';

export type ColorScheme = (typeof COLOR_SCHEMES)[keyof typeof COLOR_SCHEMES];

export type ColorSchemeOutput = {
  colorScheme: ColorScheme;
  setColorScheme: (data: ColorScheme) => void;
};
