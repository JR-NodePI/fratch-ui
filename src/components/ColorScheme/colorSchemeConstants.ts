export const COLOR_SCHEMES = Object.freeze({
  LIGHT: 'light',
  DARK: 'dark',
} as const);
