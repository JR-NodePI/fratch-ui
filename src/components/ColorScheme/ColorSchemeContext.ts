import { createContext } from 'react';

import type { ColorScheme } from './ColorScheme';
import { COLOR_SCHEMES } from './colorSchemeConstants';

const ColorSchemeContext = createContext<
  [colorScheme: ColorScheme, setColorScheme: (data: ColorScheme) => void]
>([COLOR_SCHEMES.LIGHT, () => {}]);

ColorSchemeContext.displayName = 'ColorSchemeContext';

export default ColorSchemeContext;
