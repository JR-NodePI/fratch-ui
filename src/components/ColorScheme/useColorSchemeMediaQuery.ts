import { useEffect } from 'react';

import type { ColorSchemeOutput } from './ColorScheme';
import { COLOR_SCHEMES } from './colorSchemeConstants';
import usePersistedColorScheme from './usePersistedColorScheme';

const COLOR_SCHEME_DATA_ATTR_NAME = 'data-fratch_color_scheme';
const COLOR_SCHEME_MEDIA_QUERY = '(prefers-color-scheme: dark)';

export default function useColorSchemeMediaQuery(): ColorSchemeOutput {
  const initialColorScheme = window.matchMedia(COLOR_SCHEME_MEDIA_QUERY).matches
    ? COLOR_SCHEMES.DARK
    : COLOR_SCHEMES.LIGHT;
  const [colorScheme, setColorScheme] =
    usePersistedColorScheme(initialColorScheme);

  useEffect(() => {
    document.body.setAttribute(COLOR_SCHEME_DATA_ATTR_NAME, colorScheme);
  }, [colorScheme]);

  useEffect(() => {
    const handler = (e: MediaQueryListEvent): void => {
      setColorScheme(e.matches ? COLOR_SCHEMES.DARK : COLOR_SCHEMES.LIGHT);
    };

    window
      .matchMedia(COLOR_SCHEME_MEDIA_QUERY)
      .addEventListener('change', handler);

    return () => {
      window
        .matchMedia(COLOR_SCHEME_MEDIA_QUERY)
        .removeEventListener('change', handler);
    };
  }, [setColorScheme]);

  return [colorScheme, setColorScheme];
}
