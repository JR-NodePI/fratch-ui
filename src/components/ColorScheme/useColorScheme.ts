import { useContext } from 'react';

import ColorSchemeContext from './ColorSchemeContext';

export default function useColorScheme() {
  return useContext(ColorSchemeContext);
}
