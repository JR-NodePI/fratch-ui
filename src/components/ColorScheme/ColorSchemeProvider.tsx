import { ColorSchemeProviderProps } from './ColorScheme';
import ColorSchemeContext from './ColorSchemeContext';
import useColorSchemeMediaQuery from './useColorSchemeMediaQuery';

export default function ColorSchemeProvider({
  children,
}: ColorSchemeProviderProps): JSX.Element {
  const [colorScheme, setColorScheme] = useColorSchemeMediaQuery();

  return (
    <ColorSchemeContext.Provider value={[colorScheme, setColorScheme]}>
      {children}
    </ColorSchemeContext.Provider>
  );
}
