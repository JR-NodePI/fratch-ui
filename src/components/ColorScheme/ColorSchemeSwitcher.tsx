import { useCallback } from 'react';

import { c } from '../../helpers';
import Switcher from '../Form/Switcher/Switcher';
import { IconDark, IconLight } from '../Icons';
import { ColorSchemeSwitcherProps } from './ColorScheme';
import { COLOR_SCHEMES } from './colorSchemeConstants';
import useColorScheme from './useColorScheme';

export default function ColorSchemeSwitcher({
  className,
}: ColorSchemeSwitcherProps): JSX.Element {
  const [colorSchema, setColorScheme] = useColorScheme();
  const oppositeColorSchema =
    colorSchema === COLOR_SCHEMES.LIGHT
      ? COLOR_SCHEMES.DARK
      : COLOR_SCHEMES.LIGHT;

  const handleToggleColorSchema = useCallback(
    (switchOn: boolean): void => {
      setColorScheme(switchOn ? COLOR_SCHEMES.DARK : COLOR_SCHEMES.LIGHT);
    },
    [setColorScheme]
  );

  const title = `Switch to ${oppositeColorSchema} mode`;
  const switchOn = colorSchema === COLOR_SCHEMES.DARK;
  return (
    <Switcher
      className={c(className)}
      title={title}
      labelLeft="color scheme"
      IconOff={IconLight}
      IconOn={IconDark}
      value={switchOn}
      onChange={handleToggleColorSchema}
    />
  );
}
