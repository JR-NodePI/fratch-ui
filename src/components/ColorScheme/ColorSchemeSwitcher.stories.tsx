import type { Meta, StoryObj } from '@storybook/react';

import ColorSchemeSwitcher from './ColorSchemeSwitcher';

const meta = {
  title: 'Example/ColorSchemeSwitcher',
  component: ColorSchemeSwitcher,
  argTypes: {
    className: { table: { disable: true } },
  },
} satisfies Meta<typeof ColorSchemeSwitcher>;

export default meta;

type Story = StoryObj<typeof ColorSchemeSwitcher>;

export const _ColorSchemeSwitcher: Story = {
  args: {},
};
