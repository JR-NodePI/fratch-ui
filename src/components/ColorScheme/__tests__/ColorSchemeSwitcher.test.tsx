import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { describe, expect, it, vi } from 'vitest';

import { type ColorSchemeSwitcherProps } from '../ColorScheme';
import ColorSchemeSwitcher from '../ColorSchemeSwitcher';
import useColorScheme from '../useColorScheme';

vi.mock('../useColorScheme');

describe('ColorSchemeSwitcher', () => {
  const setColorScheme = vi.fn();

  beforeEach(() => {
    (
      useColorScheme as jest.MockedFunction<typeof useColorScheme>
    ).mockReturnValue(['light', setColorScheme]);
  });

  const setup = (props: ColorSchemeSwitcherProps = {}) =>
    render(<ColorSchemeSwitcher {...props} />);

  afterEach(() => {
    vi.resetAllMocks();
    setColorScheme.mockClear();
  });

  it('should render properly', () => {
    const { container } = setup();
    expect(container).toMatchSnapshot();
  });

  it('should render with className', () => {
    const { container } = setup({
      className: 'mock_test_class_name',
    });
    expect(container).toMatchSnapshot();
  });

  it('should set a new color schema', async () => {
    setup();

    await userEvent.click(screen.getByRole('button'));

    expect(setColorScheme).toHaveBeenCalledTimes(1);
    expect(setColorScheme).toHaveBeenCalledWith('dark');
  });
});
