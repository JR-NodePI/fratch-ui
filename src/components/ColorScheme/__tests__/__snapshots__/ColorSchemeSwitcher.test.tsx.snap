// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ColorSchemeSwitcher > should render properly 1`] = `
<div>
  <label
    class="_switcher_b3fc64 _off_b3fc64"
    title="Switch to dark mode"
  >
    color scheme
    <button
      class="_switcher_button_b3fc64"
    >
      <span
        class="_switcher_button_dot_b3fc64"
      >
        <svg
          aria-hidden="true"
          class="_icon_2be8fc _primary_2be8fc _switcher_button_icon_b3fc64"
          fill="none"
          viewBox="0 0 20 20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M10 3V1m0 18v-2M5.05 5.05 3.636 3.636m12.728 12.728L14.95 14.95M3 10H1m18 0h-2M5.05 14.95l-1.414 1.414M16.364 3.636 14.95 5.05M14 10a4 4 0 1 1-8 0 4 4 0 0 1 8 0Z"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
          />
        </svg>
      </span>
    </button>
  </label>
</div>
`;

exports[`ColorSchemeSwitcher > should render with className 1`] = `
<div>
  <label
    class="_switcher_b3fc64 _off_b3fc64 mock_test_class_name"
    title="Switch to dark mode"
  >
    color scheme
    <button
      class="_switcher_button_b3fc64"
    >
      <span
        class="_switcher_button_dot_b3fc64"
      >
        <svg
          aria-hidden="true"
          class="_icon_2be8fc _primary_2be8fc _switcher_button_icon_b3fc64"
          fill="none"
          viewBox="0 0 20 20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M10 3V1m0 18v-2M5.05 5.05 3.636 3.636m12.728 12.728L14.95 14.95M3 10H1m18 0h-2M5.05 14.95l-1.414 1.414M16.364 3.636 14.95 5.05M14 10a4 4 0 1 1-8 0 4 4 0 0 1 8 0Z"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
          />
        </svg>
      </span>
    </button>
  </label>
</div>
`;
